src/sys/resource.rs

```diff
@@ -0,0 +1,230 @@
+        target_os = "android",
+        target_os = "dragonfly",
+        target_os = "bitrig",
+        target_os = "linux", // target_env != "gnu"
+    ))]{
+        use libc::{c_int, rlimit, RLIM_INFINITY};
+
+        libc_enum! {
+            #[repr(i32)]
+            pub enum Resource {
+                /// See detail of each Resource https://man7.org/linux/man-pages/man2/getrlimit.2.html
+                /// BSD specific Resource https://www.freebsd.org/cgi/man.cgi?query=setrlimit
+                #[cfg(not(any(target_os = "netbsd", target_os = "freebsd")))]
+                RLIMIT_AS,
+                RLIMIT_CORE,
+                RLIMIT_CPU,
+                RLIMIT_DATA,
+                RLIMIT_FSIZE,
+                RLIMIT_NOFILE,
+                RLIMIT_STACK,
+
+                // platform specific
+                #[cfg(target_os = "freebsd")]
+                RLIMIT_KQUEUES,
+
+                #[cfg(any(target_os = "android", target_os = "linux"))]
+                RLIMIT_LOCKS,
+
+                #[cfg(any(target_os = "android", target_os = "freebsd", target_os = "openbsd", target_os = "linux"))]
+                RLIMIT_MEMLOCK,
+
+                #[cfg(any(target_os = "android", target_os = "linux"))]
+                RLIMIT_MSGQUEUE,
+
+                #[cfg(any(target_os = "android", target_os = "linux"))]
+                RLIMIT_NICE,
+
+                #[cfg(any(target_os = "android", target_os = "freebsd", target_os = "openbsd", target_os = "linux"))]
+                RLIMIT_NPROC,
+
+                #[cfg(target_os = "freebsd")]
+                RLIMIT_NPTS,
+
+                #[cfg(any(target_os = "android", target_os = "freebsd", target_os = "openbsd", target_os = "linux"))]
+                RLIMIT_RSS,
+
+                #[cfg(any(target_os = "android", target_os = "linux"))]
+                RLIMIT_RTPRIO,
+
+                #[cfg(any(target_os = "linux"))]
+                RLIMIT_RTTIME,
+
+                #[cfg(any(target_os = "android", target_os = "linux"))]
+                RLIMIT_SIGPENDING,
+
+                #[cfg(any(target_os = "freebsd", target_os = "dragonfly"))]
+                RLIMIT_SBSIZE,
+
+                #[cfg(target_os = "freebsd")]
+                RLIMIT_SWAP,
+
+                #[cfg(target_os = "freebsd")]
+                RLIMIT_VMEM,
+            }
+        }
+    }else{
+        // unkown os
+    }
+}
+
+/// Get the current processes resource limits
+///
+/// A value of `None` indicates that there's no limit.
+///
+/// # Parameters
+///
+/// * `resource`: The [`Resource`] that we want to get the limits of.
+///
+/// # Examples
+///
+/// ```
+/// # use nix::sys::resource::{getrlimit, Resource};
+///
+/// let (soft_limit, hard_limit) = getrlimit(Resource::RLIMIT_NOFILE).unwrap();
+/// println!("current soft_limit: {:?}", soft_limit);
+/// println!("current hard_limit: {:?}", hard_limit);
+/// ```
+///
+/// # References
+///
+/// [getrlimit(2)](https://linux.die.net/man/2/getrlimit)
```

Comment:
> Rather than define `Resource` twice for different platforms, which doesn't even handle all platform-specificities, I think you should define `Resource` only once, and gate each enum item with an appropriate `#[cfg(...)]`.
> 
> I haven't reviewed the tests yet. ENOTIME.

Thank you for the suggestion. But I am not sure when enum can be both `#[repr(u32)]` and `#[repr(i32)]`, how could I do this. (The reason why we have both `u32` and `i32` is explained in one of comments below) The closest case I can think of is using something like this:

```rust
libc_enum!{
    pub enum ResourceDemo {
        #[cfg(all(target_os = "linux", target_env = "gnu"))] #[repr(u32)]
        #[cfg(any(
            target_os = "freebsd",
            target_os = "openbsd",
            target_os = "netbsd",
            target_os = "macos",
            target_os = "ios",
            target_os = "android",
            target_os = "dragonfly",
            target_os = "linux"))] #[repr(i32)]
        RLIMIT_AS,

        #[cfg(all(target_os = "linux", target_env = "gnu"))] #[repr(u32)]
        #[cfg(any(
            target_os = "freebsd",
            target_os = "openbsd",
            target_os = "netbsd",
            target_os = "macos",
            target_os = "ios",
            target_os = "android",
            target_os = "dragonfly",
            target_os = "linux"))] #[repr(i32)]
        RLIMIT_CORE,
        RLIMIT_CPU,
        RLIMIT_DATA,
        RLIMIT_FSIZE
    }
}
```
But this will have compile error for linux-musl series toolchain as both cfg will be compiled thus causing the compile error. The idea I have came out is using `cfg_if`, but if there is better ways to do this. I am happy to adopt. Thank you very much.

Reply:
Will change this in new revision